src/components/tooltip/GCMarker.js

```diff
@@ -423,80 +430,103 @@ function _treeInsert(
   let node = tree.get(component);
   if (!node) {
     // Make a new node and grow the tree in this direction.
-    node = { branches: new Map() };
+    const value = { name: '(temp)', time: 0 };
+    node = { branches: new Map(), leaf: false, value };
 
     tree.set(component, node);
   }
 
   if (path.length > 0) {
     // There are more path components.  This node should be a branch if it
     // isn't one already.
-    if (node.value) {
-      // We delete the value to change this node from a leaf to a branch.
-      delete node.value;
-    }
+    node.leaf = false;
     _treeInsert(node.branches, path, phase);
   } else {
     // Make the new node leaf node.
-    if (node.value) {
+    if (node.value.name !== '(temp)') {
       console.error(
-        'Duplicate phases in _treeInsert in MarkerTooltipContents.js'
+        `Duplicate phase ${phase.name} in _treeInsert in MarkerTooltipContents.js`
       );
       return;
     }
     node.value = phase;
+    node.leaf = true;
   }
 }
 
-function _treeGetLeaves(
-  tree: Map<string, PhaseTreeNode>
-): Array<PhaseTimeTuple> {
-  const leaves = [];
+function _treeMapNodes(
+  tree: Map<string, PhaseTreeNode>,
+  f: (node: PhaseTreeNode) => any,
+  results: Array<any> = []
+): Array<any> {
```

Comment:
Ooh, that's much better. But I'll have to look up how to say `T|undefined` or whatever -- the function can return undefined, which won't be pushed into the output array. Which I'm saying because I'm being lazy and hoping that you'll just tell me how to do it, but flow types are kinda fun so I'll probably look it up myself anyway. :-)

Thanks for the review!

Reply:
Bleh. I fixed it (by making `f` return `?T`), but I don't understand why Flow had problems with my type refinement. Specifically, later on I have:
```
const result = f(node);
if (result !== undefined) {
  results.push(result);
}
```
Flow didn't like this -- it claimed `result` could still be undefined. Nor did it like `if (typeof result !== "undefined")`. It *is* okay with `if (result != null)`, oddly enough, which will also work for undefined.